market/retrieval/piecehandler.go

```diff
@@ -58,19 +60,46 @@ func (rp *Provider) handleByPieceCid(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	setHeaders(w, pieceCid)
+	buf := make([]byte, 512)
+	n, _ := reader.Read(buf)
+	contentType := http.DetectContentType(buf[:n])
+
+	// rewind reader before sending
+	_, err = reader.Seek(0, io.SeekStart)
+	if err != nil {
+		log.Errorf("error rewinding reader for piece CID %s: %s", pieceCid, err)
+		w.WriteHeader(http.StatusInternalServerError)
+		stats.Record(ctx, remoteblockstore.HttpPieceByCid500ResponseCount.M(1))
+		return
+	}
+
+	setHeaders(w, pieceCid, pdpPiece, contentType)
 	serveContent(w, r, size, reader)
 
 	stats.Record(ctx, remoteblockstore.HttpPieceByCid200ResponseCount.M(1))
 	stats.Record(ctx, remoteblockstore.HttpPieceByCidRequestDuration.M(float64(time.Since(startTime).Milliseconds())))
 }
 
-func setHeaders(w http.ResponseWriter, pieceCid cid.Cid) {
+func setHeaders(w http.ResponseWriter, pieceCid cid.Cid, pdpPiece bool, contentType string) {
 	w.Header().Set("Vary", "Accept-Encoding")
-	etag := `"` + pieceCid.String() + `.gz"` // must be quoted
-	w.Header().Set("Etag", etag)
-	w.Header().Set("Content-Type", "application/piece")
 	w.Header().Set("Cache-Control", "public, max-age=29030400, immutable")
+	if pdpPiece {
+		w.Header().Set("Content-Type", contentType)
+		if contentType != "application/octet-stream" {
+			if exts, err := mime.ExtensionsByType(contentType); err == nil && len(exts) > 0 {
+				ext := exts[0]
+				filename := pieceCid.String() + ext
+				encoded := url.PathEscape(filename)
+
+				w.Header().Set("Content-Disposition",
+					fmt.Sprintf(`inline; filename="%s"; filename*=UTF-8''%s`, filename, encoded))
+			}
+		}
+	} else {
+		etag := `"` + pieceCid.String() + `.gz"` // must be quoted
+		w.Header().Set("Etag", etag)
+		w.Header().Set("Content-Type", "application/piece")
```

Comment:
That's not really the standard way to do mime, I'd say either the .car mime, or do octet-stream. (I know it's what we used to do, but it's still wrong)

Reply:
I would prefer to use "application/octet-stream" instead of car to align more with HTTP standards instead if Filecoin. A piece is in the end a byte array and it might not be a car file in market v2.